### PR TITLE
Add Progress Bar to docs

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Part of #7666

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @rohitgr7 @SeanNaren @kaushikb11 @awaelchli

### CODE REVIEW

**Comment by rohitgr7** [comment]
hey @kaushikb11 !

doc test failing. can you fix them for review?

File: docs/source/common/progress_bar.rst [hunk 1]
```diff
@@ -0,0 +1,138 @@
+.. testsetup:: *
+
+    from pytorch_lightning.trainer.trainer import Trainer
+
+Progress Bar
+============
+
+Lightning supports two different types of progress bars (`tqdm <https://github.com/tqdm/tqdm>`_ and `rich <https://github.com/Textualize/rich>`_). :class:`~pytorch_lightning.callbacks.TQDMProgressBar` is used by default,
+but you can override it by passing custom :class:`~pytorch_lightning.callbacks.TQDMProgressBar` or :class:`~pytorch_lightning.callbacks.RichProgressBar` to the ``callbacks`` flag of :class:`~pytorch_lightning.trainer.trainer.Trainer`.
```
Comment (ethanwharris) [inline docs/source/common/progress_bar.rst:9]:
```suggestion
but you can override it by passing a custom :class:`~pytorch_lightning.callbacks.TQDMProgressBar` or :class:`~pytorch_lightning.callbacks.RichProgressBar` to the ``callbacks`` flag of :class:`~pytorch_lightning.trainer.trainer.Trainer`.
```

File: docs/source/common/progress_bar.rst [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+    RichProgressBar
+    TQDMProgressBar
+
+-------
+
+TQDMProgressBar
+---------------
+
+It uses the `tqdm <https://github.com/tqdm/tqdm>`_ library internally and is the default progress bar used by Lightning.
```
Comment (ethanwharris) [inline docs/source/common/progress_bar.rst:29]:
```suggestion
The :class:`~pytorch_lightning.callbacks.TQDMProgressBar` uses the `tqdm <https://github.com/tqdm/tqdm>`_ library internally and is the default progress bar used by Lightning.
```

File: docs/source/common/progress_bar.rst [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    - :class:`~pytorch_lightning.callbacks.TQDMProgressBar` docs.
+    - `tqdm library <https://github.com/tqdm/tqdm>`__
+
+----------------
+
+RichProgressBar
+---------------
+
+`Rich <https://github.com/Textualize/rich>`_ is a Python library for rich text and beautiful formatting in the terminal.
+To use :class:`~pytorch_lightning.callbacks.RichProgressBar` as your progress bar do the following. First, install the package:
```
Comment (ethanwharris) [inline docs/source/common/progress_bar.rst:72]:
```suggestion
To use the :class:`~pytorch_lightning.callbacks.RichProgressBar` as your progress bar, first install the package:
```

File: docs/source/common/progress_bar.rst [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+Then configure the callback and pass it to the :class:`~pytorch_lightning.trainer.trainer.Trainer`:
+
+.. code-block:: python
+
+    from pytorch_lightning.callbacks import RichProgressBar
+
+    trainer = Trainer(callbacks=RichProgressBar())
+
+Customize the theme for your :class:`~pytorch_lightning.callbacks.RichProgressBar` by:
```
Comment (ethanwharris) [inline docs/source/common/progress_bar.rst:86]:
```suggestion
Customize the theme for your :class:`~pytorch_lightning.callbacks.RichProgressBar` like this:
```

File: docs/source/common/progress_bar.rst [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+            batch_progress="green_yellow",
+            time="grey82",
+            processing_speed="grey82",
+            metrics="grey82",
+        )
+    )
+
+    trainer = Trainer(callbacks=progress_bar)
+
+You could customize the components used within :class:`~pytorch_lightning.callbacks.RichProgressBar` with ease by overriding the
```
Comment (ethanwharris) [inline docs/source/common/progress_bar.rst:109]:
```suggestion
You can customize the components used within :class:`~pytorch_lightning.callbacks.RichProgressBar` with ease by overriding the
```

File: docs/source/common/progress_bar.rst [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+
+    class CustomRichProgressBar(RichProgressBar):
+        def configure_columns(self, trainer):
+            return [custom_column]
+
+
+    progress_bar = CustomRichProgressBar()
+
+If you wish a new progress bar to be displayed at the end of every epoch, you could enable
```
Comment (ethanwharris) [inline docs/source/common/progress_bar.rst:126]:
```suggestion
If you wish for a new progress bar to be displayed at the end of every epoch, you should enable
```

File: docs/source/common/progress_bar.rst [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+:paramref:`~pytorch_lightning.callbacks.RichProgressBar.leave` by passing ``True``
```
Comment (ethanwharris) [inline docs/source/common/progress_bar.rst:127]:
```suggestion
:paramref:`RichProgressBar.leave <pytorch_lightning.callbacks.RichProgressBar.leave>` by passing ``True``
```

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley: Some small suggestions

File: docs/source/common/progress_bar.rst [hunk 8]
```diff
@@ -0,0 +1,138 @@
+.. testsetup:: *
+
+    from pytorch_lightning.trainer.trainer import Trainer
+
+Progress Bar
+============
+
+Lightning supports two different types of progress bars (`tqdm <https://github.com/tqdm/tqdm>`_ and `rich <https://github.com/Textualize/rich>`_). :class:`~pytorch_lightning.callbacks.TQDMProgressBar` is used by default,
```
Comment (ananthsub) [inline docs/source/common/progress_bar.rst:8]: 1. should this mention the base progress bar class somewhere? 

or at this point, do we even need the base class? 
- progress tracking is done in the loop
- once https://github.com/PyTorchLightning/pytorch-lightning/issues/10982 is in, then enable/disable aren't required on the progress bar API

2. should we mention that a progress bar is automatically enabled with the Trainer, and to disable it, one should do this:`Trainer(enable_progress_bar=False)`

File: docs/source/common/progress_bar.rst [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+
+For infinite datasets, the progress bar never ends.
+
+You can update ``refresh_rate`` (rate (number of batches) at which the progress bar get updated) for :class:`~pytorch_lightning.callbacks.TQDMProgressBar` by:
+
+.. code-block:: python
+
+    from pytorch_lightning.callbacks import TQDMProgressBar
+
+    trainer = Trainer(callbacks=TQDMProgressBar(refresh_rate=10))
```
Comment (ananthsub) [inline docs/source/common/progress_bar.rst:45]: should we use list convention everywhere when specifying callbacks on the trainer? 
```suggestion
    trainer = Trainer(callbacks=[TQDMProgressBar(refresh_rate=10)])
```

File: docs/source/common/progress_bar.rst [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+    pip install rich
+
+Then configure the callback and pass it to the :class:`~pytorch_lightning.trainer.trainer.Trainer`:
+
+.. code-block:: python
+
+    from pytorch_lightning.callbacks import RichProgressBar
+
+    trainer = Trainer(callbacks=RichProgressBar())
```
Comment (ananthsub) [inline docs/source/common/progress_bar.rst:84]:
```suggestion
    trainer = Trainer(callbacks=[RichProgressBar()])
```

File: docs/source/common/progress_bar.rst [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+            progress_bar_finished="green1",
+            progress_bar_pulse="#6206E0",
+            batch_progress="green_yellow",
+            time="grey82",
+            processing_speed="grey82",
+            metrics="grey82",
+        )
+    )
+
+    trainer = Trainer(callbacks=progress_bar)
```
Comment (ananthsub) [inline docs/source/common/progress_bar.rst:107]:
```suggestion
    trainer = Trainer(callbacks=[progress_bar])
```

File: docs/source/common/progress_bar.rst [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+    from pytorch_lightning.callbacks import RichProgressBar
+
+    trainer = Trainer(callbacks=RichProgressBar())
+
+Customize the theme for your :class:`~pytorch_lightning.callbacks.RichProgressBar` like this:
+
+.. code-block:: python
+
+    from pytorch_lightning.callbacks import RichProgressBar
+    from pytorch_lightning.callbacks.progress.rich_progress import RichProgressBarTheme
```
Comment (ananthsub) [inline docs/source/common/progress_bar.rst:91]: i'm not sure if our naming is consistent here w/ other directories. should we rename this file to be
`pytorch_lightning.callbacks.progress.rich` ?

File: docs/source/common/progress_bar.rst (hunk 9 above)
Comment (carmocca) [reply docs/source/common/progress_bar.rst:45]: Why? Passing a single callback is supported.

File: docs/source/common/progress_bar.rst (hunk 12 above)
Comment (akihironitta) [reply docs/source/common/progress_bar.rst:91]: Its directory could also be renamed from `pytorch_lightning.callbacks.progress` to `pytorch_lightning.callbacks.progress_bars` which I think sounds more natural.

**Review by tchaton — APPROVED** [review]
LGTM !

File: docs/source/common/progress_bar.rst [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+
+    class LitProgressBar(TQDMProgressBar):
+        def init_validation_tqdm(self):
+            bar = super().init_validation_tqdm()
+            bar.set_description("running validation...")
+            return bar
+
+
+    bar = LitProgressBar()
+    trainer = Trainer(callbacks=[bar])
```
[the remark this answers is not in the document]
Comment (kaushikb11) [reply docs/source/common/progress_bar.rst]: Updated!

File: docs/source/common/progress_bar.rst (hunk 9 above)
Comment (kaushikb11) [reply docs/source/common/progress_bar.rst:45]: Yup.

Comment (ananthsub) [reply docs/source/common/progress_bar.rst:45]: Using the list reinforces that passing multiple callbacks is supported. This could avoid confusion for complete beginners who go through the docs and see single callbacks configured, but who aren't aware of how to chain them together. It also avoids confusion between `Trainer(callbacks=<Single class>)` having a different type than `trainer.callbacks`

File: docs/source/common/progress_bar.rst [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+    :toctree: generated
+    :nosignatures:
+    :template: classtemplate.rst
+
+    RichProgressBar
+    TQDMProgressBar
+
+-------
+
+TQDMProgressBar
```
[the remark this answers is not in the document]
Comment (carmocca) [reply docs/source/common/progress_bar.rst:26]: There's still technical limitations to do the switch right?

Keep it as it is and we can move it when it actually happens

**Review by akihironitta — COMMENTED** [review]
Seems like `:noindex:` is needed somewhere for the build to be successful. https://app.circleci.com/pipelines/github/PyTorchLightning/pytorch-lightning/51279/workflows/aa13874c-25d0-4170-8294-216da74189f5/jobs/146122?invite=true#step-103-616
```
Warning, treated as error:
/home/docs/project/pytorch_lightning/callbacks/progress/rich_progress.py:docstring of pytorch_lightning.callbacks.progress.rich_progress.RichProgressBar:1:duplicate object description of pytorch_lightning.callbacks.progress.rich_progress.RichProgressBar, other instance in common/generated/pytorch_lightning.callbacks.progress.RichProgressBar, use :noindex: for one of them
Makefile:19: recipe for target 'html' failed
```

**Review by Borda — APPROVED** [review]

**Comment by akihironitta** [comment]
Converting this PR to draft to avoid accidentally getting merged to master.

File: docs/source/common/progress_bar.rst (hunk 8 above)
Comment (kaushikb11) [reply docs/source/common/progress_bar.rst:11]: Updated!

File: docs/source/common/progress_bar.rst [hunk 15]
```diff
@@ -0,0 +1,137 @@
+.. testsetup:: *
+
+    from pytorch_lightning.trainer.trainer import Trainer
+
+Progress Bar
```
[the remark this answers is not in the document]
Comment (kaushikb11) [reply docs/source/common/progress_bar.rst:8]: Updated!
